Enhance FileSystem related APIs to be able to cache and provide the `cas::ObjectRef` for a file's contents

akyrtzi opened the pull request:
* Introduce `vfs::File::getCASContents()` which returns a `cas::ObjectRef` for filesystem implementations that are able to provide this
* Introduce `llvm::cas::createCASProvidingFileSystem()` which returns a `vfs::FileSystem` that supports `vfs::File::getCASContents()`
* Enhance `DependencyScanningWorkerFilesystem` to be able to cache and retrieve the `cas::ObjectRef` for a file's contents,
  if the underlying filesystem provided it
* Enhance `clang::FileManager::getBufferForFile()` APIs so that they can provide the `cas::ObjectRef` for a file's contents,
  if the underlying filesystem can provide it

The overall goal is to be able to efficiently retrieve a file's contents `cas::ObjectRef` during dependency scanning,
by taking advantage of `DependencyScanningWorkerFilesystem` caching mechanism and only ingesting a file once during the lifetime
of a `DependencyScanningService` instance.

akyrtzi committed:
Enhance FileSystem related APIs to be able to cache and provide the `cas::ObjectRef` for a file's contents

* Introduce `vfs::File::getCASContents()` which returns a `cas::ObjectRef` for filesystem implementations that are able to provide this
* Introduce `llvm::cas::createCASProvidingFileSystem()` which returns a `vfs::FileSystem` that supports `vfs::File::getCASContents()`
* Enhance `DependencyScanningWorkerFilesystem` to be able to cache and retrieve the `cas::ObjectRef` for a file's contents,
  if the underlying filesystem provided it
* Enhance `clang::FileManager::getBufferForFile()` APIs so that they can provide the `cas::ObjectRef` for a file's contents,
  if the underlying filesystem can provide it

The overall goal is to be able to efficiently retrieve a file's contents `cas::ObjectRef` during dependency scanning,
by taking advantage of `DependencyScanningWorkerFilesystem` caching mechanism and only ingesting a file once during the lifetime
of a `DependencyScanningService` instance.

akyrtzi commented:
@swift-ci Please test

cachemeifyoucan commented:
The high-level feedback is that we already have `CASFileSystemBase` in `llvm/CAS/ThreadSafeFileSystem.h`. We should converge those two abstractions so we don't have two.

akyrtzi commented:
> The high-level feedback is that we already have `CASFileSystemBase` in `llvm/CAS/ThreadSafeFileSystem.h`. We should converge those two abstractions so we don't have two.

Ah good point! I think `CASFileSystemBase` can be removed after these changes, I'll look into it.

cachemeifyoucan commented on clang/include/clang/Basic/FileManager.h:
```diff
@@ -272,18 +272,21 @@ class FileManager : public RefCountedBase<FileManager> {
   llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>>
   getBufferForFile(StringRef Filename, bool isVolatile = false,
-                   bool RequiresNullTerminator = true) {
+                   bool RequiresNullTerminator = true,
```
I am not sure if I like that you get the ObjectRef when you getBuffer. Maybe use a new API that returns `ErrorOr<ObjectRef>`, or add a new API that retrieve both with a different name if that is what we always doing.

akyrtzi replied:
The general idea with `getBufferForFile()` returning both is that after you open a file you'd get both the buffer and the `ObjectRef` for the buffer for "free", so you'd avoid 2 lookups if you need both of them.

I can see the usefulness of adding a new API to return an `ObjectRef` only, but do you have a concern with adding the optional parameter to `getBufferForFile()`? If I add an extra set of `getBufferAndObjectRefForFile()` functions it will be essentially the same signatures as the existing `FileManager::getBufferForFile()` functions but with an added parameter, in which case it seems natural to just add it as an optional parameter.

cachemeifyoucan replied:
Yes, I can understand that, which is why I made two proposals depending on your use case. But now with a second thought, I wonder if this is the correct way to model CASFileSystem.

Maybe CASID or ObjectRef should not be the product after opening the file, but more like `status` that you can get without opening the file? Then you can cache the ObjectRef in the FileSystemStatCache if needed so using another call to get ObjectRef is not expensive.

akyrtzi replied:
> Maybe CASID or ObjectRef should not be the product after opening the file, but more like status that you can get without opening the file?

I'm a bit concerned that "ingestion to get ObjectRef", whenever it happens for `CASProvidingFileSystem`, would do relatively significant work (open the file, get the buffer and ingest it), while "get status" is generally expected to be way more lightweight than that, so there'd be a disconnect that I'd prefer to avoid.

There's possibility in the future for a `vfs::FileSystem` implementation where getting `ObjectRef` is as cheap as "get status" but we could re-evaluate if the FileSystem/FileManager APIs would benefit from another change if this comes to be.

cachemeifyoucan replied:
I guess we are on the same page? My point was that we should make `getCASContentsForFile` cheap instead of getting it for `free` when you open it for reading. Currently, it requires hashing so it should just be a field stored in `CASProvidingFile`.

Also `getCASContentsForFile` is not a good name. What you get is a CAS ObjectRef, not the content of the file.

akyrtzi replied:
> I guess we are on the same page? My point was that we should make getCASContentsForFile cheap instead of getting it for free when you open it for reading. Currently, it requires hashing so it should just be a field stored in CASProvidingFile.

You suggested "more like `status` that you can get without opening the file" and the thing I wanted to convey is that it shouldn't be the case that just by calling `vfs::FileSystem::status()` it should give you the `ObjectRef` for the file buffer, because the act of getting the `ObjectRef` opens the file and reads its buffer, which is way more expensive than what `status()` is expected to do.

Also it's not clear to me what specific source change you are suggesting, could you elaborate? You suggested "should just be a field stored in CASProvidingFile" and "instead of getting it for free when you open it for reading" but these are contradictory because to get the `CASProvidingFile` you need to call `vfs::FileSystem::openFileForRead()` so it is still the case of "getting it for free when you open it for reading" and it won't be "more like `status`".

> Also getCASContentsForFile is not a good name. What you get is a CAS ObjectRef, not the content of the file.

How about `getObjectRefForContentsOfFile`?

cachemeifyoucan replied:
I am suggesting `CASProvidingFile` to have a field `ObjectRef` so it would be cheap when you try to get ObjectRef once you have the file created. Yes, it currently requires openFile but that is implementation details and it might not be needed if the file system is CAS backed. The point is:
* You should be able to get ObjectRef without openFile (from API's point of view)
* It should not be more expensive to get ObejctRef and openFile in two API calls.

> How about getObjectRefForContentsOfFile?

That seems long. `getObjectRefForContent` is probably good enough.

akyrtzi replied:
> You should be able to get ObjectRef without openFile (from API's point of view)

So this is the `vfs::FileSystem::getCASContentsForFile()` function in this PR, right?

> It should not be more expensive to get ObejctRef and openFile in two API calls.

I'd prefer to keep the implementation of `CASProvidingFileSystem` as simple as possible for now (instead of have it do its own ad-hoc caching for its queries) since its primary use-case is to be the underlying file-system of `DependencyScanningWorkerFilesystem` which already does its own caching, and any other CAS-related file-systems can implement the `vfs::File::getCASContents()` function with their own data/caches.

Later on, if there is a need, we can add caching on top or introduce something like `CachingCASProvidingFileSystem` or just compose `DependencyScanningWorkerFilesystem`+`CASProvidingFileSystem` for a file-system that both provides `ObjectRef` for contents and also caches  openFile calls.

> That seems long. getObjectRefForContent is probably good enough.

SGTM 👍

akyrtzi committed:
Add `vfs::FileSystem::getCASContentsForFile()` and `clang::FileManager::getCASContentsForFile()` functions

These are intended for callers that don't need to get both the file buffer and the buffer's `cas::ObjectRef`.

akyrtzi committed:
[llvm/CAS] Remove `CASFileSystemBase`

It's been superseeded via the `vfs::File::getCASContents()` API addition.

akyrtzi commented:
@swift-ci Please test

akyrtzi commented:
@cachemeifyoucan I've added API to get the `ObjectRef` only (https://github.com/apple/llvm-project/pull/5004/commits/4a31089ddaaa3d2e9c89c117c2cf245060c9c259) and removed `CASFileSystemBase` (https://github.com/apple/llvm-project/pull/5004/commits/ccf115d18cf69b0aaf510cc0f0dc09ee2fcd4f79).

akyrtzi committed:
Rename `File::getCASContents()` -> `File::getObjectRefForContent()` and `FileSystem::getCASContentsForFile()` -> `FileSystem::getObjectRefForFileContent()`, NFC

Kept "File" in `FileSystem::getObjectRefForFileContent()` to make it more clear and consistent with `FileSystem::getBufferForFile()`.

akyrtzi commented:
@swift-ci Please test

cachemeifyoucan reviewed (approved):
LGTM
